libminifi/src/Connection.cpp

```diff
@@ -224,11 +216,11 @@ std::shared_ptr<core::FlowFile> Connection::poll(std::set<std::shared_ptr<core::
 }
 
 void Connection::drain(bool delete_permanently) {
-  std::lock_guard<std::mutex> lock(mutex_);
+  auto lockedQueue = queue_.lock();
```

Comment:
A better pattern would be just to lock the queue until you move the data out of it.

Reply:
You mean only lock it before calling empty and unlock after pop_front? That's not an equivalent refactor. I don't know how costly the operations are after pop-ing the element but locking in a loop can definitely degrade the performance.